### PR TITLE
Runtime: Refactor builtins module

### PR DESCRIPTION
**Pull request by buffalojoec**
This is chunk 1/7 of the broken-up PR #79.

#### Problem
As we prepare to migrate builtin programs to Core BPF, we need to restructure the runtime's `builtins` module so we can add the migration code and abstract it away behind types such as `BuiltinPrototype`.

In order to minimize swelling of `bank.rs`, it makes sense to refactor `builtins` into a submodule directory, where migration code can live. This migration code can be called from the `BuiltinPrototype` objects that the bank iterates over.

#### Summary of Changes
- Refactor the `builtins` module to be a submodule directory.
- Refactor the `feature_id` field on the `BuiltinPrototype` to be `enable_feature_id`, since that's what this field is responsible for.
- Add a prototype and static list for "stateless builtins" - builtin programs with no on-chain account (reserved keys).

### CODE REVIEW

File: runtime/src/bank/builtins/prototypes.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        }
+    }
+}
+
+/// Transitions of stateless built-in programs at epoch boundaries when
+/// features are activated.
+/// These are built-in programs that don't actually exist, but their address
+/// is reserved.
+#[derive(Debug)]
+pub struct StatelessBuiltinPrototype {
```
Comment (CriesofCarrots) [inline runtime/src/bank/builtins/prototypes.rs:45]: Is there value in implementing `Debug` for this prototype too?

File: runtime/src/bank/builtins/mod.rs [hunk 2]
```diff
@@ -1,110 +1,78 @@
...
-        });
-        Self {
-            feature_id: None,
-            program_id: Pubkey::default(),
-            name: "",
-            entrypoint: MockBuiltin::vm,
-        }
-    }
-}
+pub use prototypes::{BuiltinPrototype, StatelessBuiltinPrototype};
```
Comment (CriesofCarrots) [inline runtime/src/bank/builtins/mod.rs:3]: I'm a bit ambivalent about this re-export. What is your reasoning for it?

**Review by CriesofCarrots — COMMENTED** [review]
Looks good, just a couple small questions.

File: runtime/src/bank/builtins/mod.rs (hunk 2 above)
Comment (buffalojoec) [reply runtime/src/bank/builtins/mod.rs:3]: Merely to preserve existing imports on other files, nothing more. I can cut this and use `runtime::builtins::prototypes::_` in those areas. I'm not sure how many files that changes, but should only be a few across bank, snapshots, and ledger.

File: runtime/src/bank/builtins/prototypes.rs (hunk 1 above)
Comment (buffalojoec) [reply runtime/src/bank/builtins/prototypes.rs:45]: Just to be consistent with the other prototype definition. We can forgo it for now until we need it, as I'm pretty sure it's unused in anything I laid down, including tests (ie. `assert_eq!`).

Comment (CriesofCarrots) [reply runtime/src/bank/builtins/prototypes.rs:45]: Oh, I'm sorry, I was looking for a custom `Debug` implementation, and missed the derive 🤦‍♀️
This works!

File: runtime/src/bank/builtins/mod.rs (hunk 2 above)
Comment (CriesofCarrots) [reply runtime/src/bank/builtins/mod.rs:3]: Okay, got it, thanks for the answer! I don't feel very strongly, so we can go with it as is.

**Review by CriesofCarrots — APPROVED** [review]

**Pull request merged**